### PR TITLE
Created public get_topic_list() function for use in other scripts.

### PR DESCRIPTION
**Pull request by EliteMasterEric**
Resolves issue #946.

The only function in rostopic for listing the available topics is built exclusively for the command line interface; this adds a public function for other applications that want a list of topics.

Usage:
```
import rostopic
publishers, subscribers = rostopic.get_topic_list()
for topic, type, node_count in publishers:
    print("Publisher: %s, [%s] (%d nodes)" % (topic, type, node_count))
for topic, type, node_count in subscribers:
    print("Subscriber: %s, [%s] (%d nodes)" % (topic, type, node_count))
```

### CODE REVIEW

**Comment by dirk-thomas** [comment]
After extracting the print-agnostic part of the existing functionality as you did in the current patch please also update the existing code to use this new function (rather than duplicating this logic). Please also consider adding a test for the new function.

**Comment by EliteMasterEric** [comment]
@dirk-thomas Changes have been made, removing redundancy and adding a test to the rostopic test script.

File: tools/rostopic/src/rostopic/__init__.py [hunk 1]
```diff
@@ -1155,6 +1155,31 @@ def topic_type(t, topic_types):
...
+            return matches[0]
+        return 'unknown type'
+
+    # Return an array of tuples; (<topic>, <type>, <node_count>)
+    master = rosgraph.Master('/rostopic')
+    state = master.getSystemState()
+    topic_types = _master_get_topic_types(master)
+    
+    pubs, subs, _ = state
+    publishers = []
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: This line is overwriting the function parameter.

Same below for `subscribers`.

File: tools/rostopic/src/rostopic/__init__.py [hunk 2]
```diff
@@ -1155,6 +1155,31 @@ def topic_type(t, topic_types):
...
+    if publishers:
+        for topic, nodes in pubs:
+            publishers.append(topic, topic_type(topic, topic_types), len(nodes))
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: `list.append()` only accepts a single argument.

Same below.

File: tools/rostopic/src/rostopic/__init__.py [hunk 3]
```diff
@@ -1202,14 +1227,7 @@ def _rostopic_list(topic, verbose=False,
...
     try:
-        state = master.getSystemState()
-
-        pubs, subs, _ = state
-        if topic:
-            # filter based on topic
-            topic_ns = rosgraph.names.make_global_ns(topic)        
-            subs = (x for x in subs if x[0] == topic or x[0].startswith(topic_ns))
-            pubs = (x for x in pubs if x[0] == topic or x[0].startswith(topic_ns))
+        pubs, subs = get_topic_list()
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py:1222]: The new function returns different information than `getSystemState` before. I doubt that all following code continues to work as is.

**Review by dirk-thomas — COMMENTED** [review]
Please make sure to run all code parts you have changed in all possible variations to make sure the patch actually works (which is currently not the case).

**Pull request closed without merging**